### DISCUSSION TITLE
How to serve NextJS static files through custom Express server?

### DISCUSSION THREAD

**Discussion by iammohitsakhuja** [vercel/next.js#14532] [Help] [answered]
My use case is that I need to run a somewhat expensive middleware for the routes that are served through `pages`. 

I have a custom server whose code kinda looks like this (unnecessary items omitted for brevity):

```typescript
const handle = app.getRequestHandler()

const bootstrap = async () => {
    await app.prepare()

    const server = express()

    server.use(somewhatExpensiveMiddleware)

    server.use(customRouter) // has logic for calling `app.render(req, res, pathname)` accordingly

    server.all('*', (req, res, next) => {
         return handle(req, res)
    })
}
```

That middleware needs to run on which request which is for whatever is in the `pages` directory. However with this setup, the `handle` method provided by Next.js is the one that serves that static asset. Is there a way through which I can place the handling of that static asset earlier in my middleware chain - for both dev and production environments, without breaking any functionality?

**Note:** the static asset could be both: either placed in `public` directory or the compiled `css`, `js` by Next.js which is served through `.next`.

Thanks in advance! 🙏

**Comment by arso1er** [comment, marked as answer]
Hi,

To serve static assets in public folder with express, require `path` module (no need to install it first) and use `server.use(express.static(path.join(__dirname, 'public')));`.

**Reply by iammohitsakhuja** [reply]
Ahhh, I was under the impression that doing this might cause problem with Next.js' own handling of serving static files. Seems like it does not cause any such problem. So, thank you! 😄 

Here's the final approach I went with for anyone that comes looking for this:

```javascript
    const server = express()

    // Serve static files.
    server.use(express.static(path.join(__dirname, '../public')))
    server.use('/_next', express.static(path.join(__dirname, '../.next')))

    server.use(somewhatExpensiveMiddleware)
```

**Reply by timbicker** [reply]
@arso1er @mohitsakhuja 
Does this allow me to serve statically generated pages with a custom server?

I am confused about this because [in the doc](https://nextjs.org/docs/advanced-features/custom-server) it says that a custom server will remove features like Automatic Static Optimization.

**Comment by muhammad-ali-abbasi** [comment]
this is how i solved this problem by just making route to "/images" to serve static images.
const express = require("express");
const next = require("next");
const path = require("path");
const bodyParser = require("body-parser");

// const dev = process.env.NODE_ENV !== "production";
const app = next({ dev: false });
const handle = app.getRequestHandler();
const port = process.env.PORT || 3004;

app.prepare().then(() => {
  const server = express();
  server.use("/images", express.static(path.join(__dirname, "public/images")));
  server.use(bodyParser.json({ limit: "10mb", extended: true }));
  server.use(bodyParser.urlencoded({ limit: "10mb", extended: true }));
  server.all("*", (req, res) => {
    return handle(req, res);
  });

  server.listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready  on http://localhost:${port}`);
  });
});

**Comment by dipak7587** [comment]
Hi,
Create a public folder by default public treats as static files

Thanks & Regards,
Dipak Gupta
+971562833180, +918708383476
Full Stack Developer
